File: SecureChat_android/app/src/main/java/com/chaosinmotion/securechat/encapsulation/SCMessageObject.java
/*
 * Copyright (c) 2016. William Edward Woody
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>
 *
 */

package com.chaosinmotion.securechat.encapsulation;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.text.Layout;
import android.text.StaticLayout;
import android.text.TextPaint;

import com.chaosinmotion.securechat.utils.JPEGImage;
import com.chaosinmotion.securechat.utils.Size;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;

/**
 * Represents a message in our system. Allows us to encapsulate different types.
 * Created by woody on 4/26/16.
 */
public class SCMessageObject
{
    /*
     *  TODO: Different types
     */
    private String message;
	private JPEGImage image;

    /**
     * Construct using raw serialized data
     * @param data
     */
    public SCMessageObject(byte[] data)
    {
        try {
	        if ((data.length > 2) && (data[0] == 0x01)) {
		        if (data[1] == 0x00) {
			        image = new JPEGImage(data,2,data.length-2);
		        } else if (data[1] == 0x01) {
			        message = new String(data,2,data.length-2,"UTF-8");
		        }
	        } else {
		        message = new String(data, "UTF-8");
	        }
        }
        catch (UnsupportedEncodingException e) {
            // Should never happen
        }
    }

    /**
     * Construct a new message object
     * @param msg
     */
    public SCMessageObject(String msg)
    {
        message = msg;
    }

	/**
	 * Construct new message object from bitmap
	 * @param bmap
	 */
    public SCMessageObject(Bitmap bmap)
    {
	    image = new JPEGImage(bmap);
    }

    /**
     * Serialize the message for transmitting
     * @return
     */
    public byte[] dataFromMessage()
    {
        try {
	        if (message != null) {
		        if ((message.length() >= 1) && (message.charAt(0) == 0x01)) {
			        // Must encapsulate. Normally should not happen
			        ByteArrayOutputStream baos = new ByteArrayOutputStream();
			        baos.write(0x01);
			        baos.write(0x01);
			        baos.write(message.getBytes("UTF-8"));
			        return baos.toByteArray();
		        } else {
			        return message.getBytes("UTF-8");
		        }
	        } else if (image != null) {
		        ByteArrayOutputStream baos = new ByteArrayOutputStream();
		        baos.write(0x01);
		        baos.write(0x00);
		        baos.write(image.getBytes());
		        return baos.toByteArray();
	        } else {
		        return null;
	        }
        }
        catch (IOException e) {
            return new byte[0];
        }
    }

    /**
     * Return a summary of the message suitable for displaying on the home page
     * @return
     */
    public String getSummaryMessageText()
    {
	    if (image != null) {
		    return "(photo)";   // TODO: what should this return?
	    } else if (message != null) {
		    return message;
	    } else {
		    return null;
	    }
    }

	/**
	 * Calculate the maximum width of this object.
	 * @return
	 */
	public int maximumWidth(Paint paint)
	{
		if (image != null) {
			return image.getWidth();
		} else if (message != null) {
			return (int)paint.measureText(message);
		} else {
			return 40;      // some random value.
		}
	}

	/**
	 * Calculate the size given the width
	 * @param width
	 * @return
	 */
	public Size sizeForWidth(TextPaint paint, int width)
	{
		if (image != null) {
			return image.sizeForWidth(width);
		} else if (message != null) {
			StaticLayout layout = new StaticLayout(message,paint,width, Layout.Alignment.ALIGN_NORMAL,1,0,true);
			int height = layout.getHeight();

			int maxWidth = 0;
			int i,len = layout.getLineCount();
			for (i = 0; i < len; ++i) {
				int w = (int)Math.ceil(layout.getLineWidth(i));
				if (w > maxWidth) {
					maxWidth = w;
				}
			}
			return new Size(maxWidth,height);

		} else {
			return new Size(width,width);
		}
	}

	/**
	 * Draw within the rectangle provided.
	 * @param canvas
	 * @param left
	 * @param top
	 * @param right
	 * @param bottom
	 * @param color
	 */
	public void drawWithRect(Canvas canvas, TextPaint paint, int left, int top, int right, int bottom)
	{
		if (image != null) {
			image.drawInRect(canvas,paint,left,top,right,bottom);
		} else if (message != null) {
			StaticLayout layout = new StaticLayout(message,paint,right-left, Layout.Alignment.ALIGN_NORMAL,1,0,true);
			canvas.save();
			canvas.translate(left,top);
			layout.draw(canvas);
			canvas.restore();
		}
	}
}

File: SecureChat_iOS/SecureChat/Encapsulation/SCMessageObject.m
//
//  SCMessageObject.m
//  SecureChat
//
//  Created by William Woody on 4/26/16.
//  Copyright © 2016 William Edward Woody. All rights reserved.
//

/*	SecureChat: A secure chat system which permits secure communications 
 *  between iOS devices and a back-end server.
 *
 *	Copyright © 2016 by William Edward Woody
 *
 *	This program is free software: you can redistribute it and/or modify it 
 *	under the terms of the GNU General Public License as published by the 
 *	Free Software Foundation, either version 3 of the License, or (at your 
 *	option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but 
 *	WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 *	or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License 
 *	for more details.
 *
 *	You should have received a copy of the GNU General Public License along 
 *	with this program. If not, see <http://www.gnu.org/licenses/>
 */


#import "SCMessageObject.h"
#import "UIImage+SCResizeImage.h"

@interface SCMessageObject ()
@property (strong) UIImage *image;
@property (copy) NSString *message;
@end

@implementation SCMessageObject

- (id)initWithData:(NSData *)data
{
	if (nil != (self = [super init])) {
		NSUInteger len = [data length];
		uint8_t *bytes = (uint8_t *)[data bytes];

		self.image = nil;
		self.message = nil;

		if ((len > 2) && (bytes[0] == 0x01)) {
			/*
			 *	Escape
			 */

			NSData *d = [NSData dataWithBytesNoCopy:bytes+2 length:len-2 freeWhenDone:NO];
			if (bytes[1] == 0x00) {
				// 0x01 0x00: JPEG
				self.image = [UIImage imageWithData:d];

			} else if (bytes[1] == 0x01) {
				// 0x01 0x01: Message
				self.message = [[NSString alloc] initWithBytes:bytes length:len encoding:NSUTF8StringEncoding];

			}
		} else {
			self.message = [[NSString alloc] initWithBytes:bytes length:len encoding:NSUTF8StringEncoding];
		}
	}
	return self;
}

// Create a message to be sent
- (id)initWithString:(NSString *)msg
{
	if (nil != (self = [super init])) {
		self.message = msg;
	}
	return self;
}

// Create a message to be sent
- (id)initWithImage:(UIImage *)image
{
	if (nil != (self = [super init])) {
		self.image = image;
	}
	return self;
}


// Encode this message as a data blob
- (NSData *)dataFromMessage
{
	NSData *data;
	uint8_t b[2];

	if (self.message) {
		data = [self.message dataUsingEncoding:NSUTF8StringEncoding];
		if (data.length < 1) return data;

		[data getBytes:b range:NSMakeRange(0, 1)];
		if (b[0] != 1) return data;

		b[0] = 1;
		b[1] = 1;
	} else if (self.image) {
		data = UIImageJPEGRepresentation(self.image, 0.75);

		b[0] = 1;
		b[1] = 0;
	} else {
		return nil;
	}

	NSMutableData *ret = [[NSMutableData alloc] initWithCapacity:data.length + 2];

	[ret appendBytes:b length:2];
	[ret appendData:data];

	return ret;
}

// Summary view content
- (NSString *)summaryMessageText
{
	if (self.message) {
		return self.message;
	} else if (self.image) {
		return NSLocalizedString(@"(photo)", @"Photo filler");
	} else {
		return nil;
	}
}

// Bubble rendering management
- (CGSize)sizeForWidth:(CGFloat)width
{
	if (self.message) {
		NSDictionary *d = @{ NSFontAttributeName: [UIFont systemFontOfSize: UIFont.labelFontSize] };
		CGRect r = [self.message boundingRectWithSize:CGSizeMake(width, 9999) options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingTruncatesLastVisibleLine attributes:d context:nil];
		return r.size;
	} else if (self.image) {
		return [self.image sizeForWidth:width];
	} else {
		// TODO: Size?
		return CGSizeMake(44, 22);
	}
}

- (void)drawWithRect:(CGRect)rect withTextColor:(NSString *)txtColor
{
	if (self.message) {
		NSDictionary *d = @{ NSFontAttributeName: [UIFont systemFontOfSize: UIFont.labelFontSize],
							 NSForegroundColorAttributeName: txtColor };

		[self.message drawInRect:rect withAttributes: d];
	} else if (self.image) {
		UIImage *i = [self.image resizeToSize:rect.size];
		CGSize size = i.size;

		rect.origin.x += (rect.size.width - size.width)/2;
		rect.origin.y += (rect.size.height - size.height)/2;
		rect.size = size;

		[i drawInRect:rect];
	}
}

@end
